### PR TITLE
Let ci-cleaner build docker image, avoid building on cronjob execution

### PR DESCRIPTION
**Pull request by marians**
ci-cleaner is running every hour as a cronjob. Currently it builds the ci-cleaner binary every time before execution.

To save CI resources, this PR lets us build a docker image on merge commits. This image is executed in cronjob runs.

To allow building using `architect`, this PR also adds a `ci-cleaner version` command, which architect expects to exist.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -31,20 +43,29 @@ jobs:
...
+        command: |
+            if [ "${CIRCLE_USERNAME}" != "" ]; then
+                wget -q $(curl -sS -H "Authorization: token $RELEASE_TOKEN" https://api.github.com/repos/giantswarm/architect/releases/latest | grep browser_download_url | head -n 1 | cut -d '"' -f 4)
+                chmod +x ./architect
+                ./architect version
+            fi
+
+    - run:
+        name: architect build (except when in cronjob execution)
+        command: test -x ./architect && ./architect build
```
Comment (JosephSalisbury) [inline .circleci/config.yml]: Why do we need the `test` here?

File: .circleci/config.yml [hunk 2]
```diff
@@ -3,25 +3,37 @@ cleanAWS: &cleanAWS
...
     machine: true
     steps:
     - checkout
 
-    - attach_workspace:
-        at: .
-
-    - run: ./ci-cleaner azure --client-id ${AZURE_CLIENT_ID} --client-secret ${AZURE_CLIENT_SECRET} --tenant-id ${AZURE_TENANT_ID} --subscription-id ${AZURE_SUBSCRIPTION_ID} --location ${AZURE_LOCATION}
-
+    - run:
```
Comment (JosephSalisbury) [inline .circleci/config.yml]: For my curiosity, why don't we need to setup AWS credentials?

File: .circleci/config.yml [hunk 3]
```diff
@@ -31,20 +43,29 @@ jobs:
...
+            fi
+
+    - run:
+        name: architect build (except when in cronjob execution)
+        command: test -x ./architect && ./architect build
+
+    - deploy:
+        name: architect deploy (master only, except when in cronjob execution)
+        command: |
+            if [ "${CIRCLE_BRANCH}" == "master" ]; then
```
Comment (JosephSalisbury) [inline .circleci/config.yml]: Do we need to test for cronjob execution here?

File: .circleci/config.yml [hunk 4]
```diff
@@ -71,7 +92,7 @@ workflows:
   cleanup:
     triggers:
       - schedule:
-          cron: "0 * * * *"
+          cron: "21 * * * *"
```
Comment (JosephSalisbury) [inline .circleci/config.yml]: Why the change?

File: .circleci/config.yml (hunk 1 above)
Comment (marians) [reply .circleci/config.yml]: In the step before, we only download architect if in the push/PR context (`CIRCLE_USERNAME != ""`), not in the cronjob context.

I figured the easiest way to pull this through to subsequent steps would be to use the existence of the architect binary.

File: .circleci/config.yml (hunk 2 above)
Comment (marians) [reply .circleci/config.yml]: We do, but I reformatted the step to give it a proper name

Comment (marians) [reply .circleci/config.yml]: For that, look at step 2 of `cleanAWS`

File: .circleci/config.yml (hunk 3 above)
Comment (marians) [reply .circleci/config.yml]: If architect is in place (test -x ./architect) we are not in a cronjob

File: .circleci/config.yml (hunk 4 above)
Comment (marians) [reply .circleci/config.yml]: because of all cronjobs in the world would execute at :00, we'd go to hell ;-)

File: .circleci/config.yml (hunk 2 above)
Comment (JosephSalisbury) [reply .circleci/config.yml]: Sorry, I meant _azure_ credentials

File: .circleci/config.yml (hunk 3 above)
Comment (JosephSalisbury) [reply .circleci/config.yml]: I get this now

File: .circleci/config.yml (hunk 4 above)
Comment (JosephSalisbury) [reply .circleci/config.yml]: Okay, so just arbitrary time change. Good enough

File: .circleci/config.yml (hunk 2 above)
Comment (marians) [reply .circleci/config.yml]: All credentials are available in Circle's environment variables. For Azure, we can just make use of them in flags. For AWS, we distinguish between control plane account and tenant cluster account. See the `TYPE` variable being set down below. That's what `aws_credentials.sh` takes care of.

**Comment by marians** [comment]
@MarcelMue Can you tell me if we still need these `requires: ['build']` entries? Like https://github.com/giantswarm/ci-cleaner/pull/31/files#diff-1d37e48f9ceff6d8030570cd36286a61R104

**Comment by MarcelMue** [comment]
If you remove the `requires: [build]` then the merge to master will run cleanup before the version you are merging will be pushed to quay. So on merge `cleanup` would use an older version.

**Comment by marians** [comment]
OK. Then let's leave it this way. Thanks!
